File: day23.py
debug = False

def rotate(l, n):
  return l[n:] + l[:n]

def run(input):
  cups = [int(c) for c in input]
  min_cup = 1
  max_cup = len(cups)

  cups.extend(range(max_cup+1, 1000001))
  max_cup = len(cups)
  debug and print(len(cups))

  # map of number -> next number. essentially a linked list
  cup_map = {}

  for i in range(len(cups)):
    cup = cups[i]
    next_cup = None
    if i < len(cups) - 1:
      next_cup = cups[i+1]
    else:
      next_cup = cups[0]
    cup_map[cup] = next_cup
  debug and print(cup_map)

  cur = cups[0]

  for i in range(10000000):
    debug and print(f'{i} BEGIN ({cur}): {cup_map}')
    n = cup_map[cur]

    picked_up = []
    for j in range(3):
      picked_up.append(n)
      n = cup_map[n]
    cup_map[cur] = n

    debug and print(f'{i} CUR: {cur} PICKED UP: {picked_up}')
    dest = cur - 1
    while dest in picked_up or dest < min_cup:
      dest -= 1
      if dest < min_cup:
        dest = max_cup

    dest_n = cup_map[dest]

    for p in picked_up:
      cup_map[dest] = p
      dest = p

    cup_map[dest] = dest_n
    cur = n

    if (i+1) % 100000 == 0:
      print(f'iteration {i+1}')

  el = 1
  if debug:
    for i in range(len(cups)-1):
      print(f'{cup_map[el]}', end='')
      el = cup_map[el]
    print()
  num = cup_map[1]
  debug and print(cup_map)
  print(num, cup_map[num], cup_map[num] * num)


# run('389125467') # example
run('643719258') # problem

File: 2020/day23.go
package main

import "fmt"

func rotate(l []int, n int) []int {
  return append(l[n:], l[:n]...)
}

func contains(s []int, e int) bool {
	for _, a := range s {
    if a == e {
      return true
    }
  }
  return false
}

func index(s []int, e int) int {
  for p, v := range s {
    if (v == e) {
      return p
    }
  }
  return -1
}


func run(cups []int) {
	minCup := 1
	maxCup := len(cups)
	fmt.Println("run")
	pos := 0
	for i := 0; i < 10000000; i++ {
    cur := cups[pos]
    // # picked_up = [cups.pop(1) for i in range(3)]
    pickedUp := []int{}
    for j := 0; j < 3; j++ {
    	pickedUp = append(pickedUp, cups[(pos + j + 1)%len(cups)])
    }
    dest := cur - 1
    for ; contains(pickedUp, dest) || dest < minCup; {
    	dest -= 1
    	if dest < minCup {
	    	dest = maxCup
    	}
    }
    destIdx := index(cups, dest)
    pickupIdx := (pos + 1) % len(cups)
    pickupEndIdx := (pos + 4) % len(cups)
    newCups := []int{}
    if pickupEndIdx < pickupIdx || pickupIdx == 0 {
      newCups = append(newCups, cups[pickupEndIdx:destIdx+1]...)
      newCups = append(newCups, pickedUp...)
      newCups = append(newCups, cups[destIdx+1:pos+1]...)
    } else if destIdx > pos {
      newCups = append(newCups, cups[:pos+1]...)
      newCups = append(newCups, cups[pos+4:destIdx+1]...)
      newCups = append(newCups, pickedUp...)
      newCups = append(newCups, cups[destIdx+1:]...)
    } else {
      newCups = append(newCups, cups[:destIdx+1]...)
      newCups = append(newCups, pickedUp...)
      newCups = append(newCups, cups[destIdx+1:pos+1]...)
      newCups = append(newCups, cups[pos+4:]...)
    }
    cups = newCups
    pos = (index(cups, cur) + 1) % len(cups)

    if (i+1) % 10000 == 0 {
      fmt.Printf("iteration %d\n", i+1)
    }

  }
  // fmt.Println("Final cups:")
  // for i := 0; i < len(cups); i++ {
  // 	fmt.Printf("%d, ", cups[i])
  // }
  fmt.Println("\n\nAfter 1:")
  fmt.Println(cups[(index(cups, 1) + 1) % len(cups)])
  fmt.Println(cups[(index(cups, 1) + 2) % len(cups)])
}

func main() {
	// input := []int{3,8,9,1,2,5,4,6,7}
	input := []int{6,4,3,7,1,9,2,5,8}

	for i := len(input)+1; i <= 1000000; i++ {
		input = append(input, i)
	}

  run(input)
}
